### PR TITLE
Added numexpr backend. Supersedes #3937

### PR DESCRIPTION
**Pull request by mihirparadkar**
Linked: #3937 
I opened a new pull request because I wanted to merge a different branch.

### CODE REVIEW

File: chainer/backends/numexpr.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+import numpy
+
+_numexpr_version = None
+_error = None
+
+try:
+    import numexpr  # NOQA
+    _numexpr_version = 0
+    numexpr_dtypes = {
```
Comment (niboshi) [inline chainer/backends/numexpr.py]: Prefix this variable with underscore (`_`).

**Comment by mihirparadkar** [comment]
It's WIP in the same sense as the IDeep backend, in that it doesn't support most of the activations and optimizers. However, as you said, it does work for the places it is implemented. I'll fix those minor things shortly.

**Comment by niboshi** [comment]
"WIP" in the PR title means "this PR is not ready for merge". In that sense, iDeep integration is no longer WIP (because it's already merged).

For this PR, it's OK to merge with minimal support and work on more implementation after that.

**Comment by mihirparadkar** [comment]
In that case, "witness the firepower of this fully armed and operational numexpr backend" :)

**Pull request closed without merging**